Speed up review pages

dpoirier opened the pull request:
E.g. /2016/reviews/section/talks/

Pre-changes: 4396 queries in 1241.01ms
Post-changes:  21 queries in 45.32ms

dpoirier commented:
Fixes: https://github.com/PyCon/pycon/issues/346

rebecca-caktus commented on symposion/reviews/views.py:
```diff
@@ -42,38 +42,64 @@ def proposals_generator(request, queryset, user_pk=None, check_speaker=True):
-        yield obj
+        try:
+            obj.category = category_by_id[obj.category_id]
+        except AttributeError:
```
When does this raise an attribute error?

dpoirier replied:
There are some proposal models that don't have a category field.

rebecca-caktus commented on symposion/reviews/views.py:
```diff
@@ -112,9 +138,9 @@ def review_section(request, section_slug, assigned=False):
+    queryset = queryset.select_related("result", "kind", "speaker__user", "tags").select_subclasses()
 
-    proposals = proposals_generator(request, queryset)
+    proposals = list(proposals_generator(request, queryset))
```
Two of three places that call `proposals_generator` use it as a list. The third is similar to the others so I think that it could be used that way as well. Perhaps the function can return a list?

dpoirier replied:
It actually did end up returning a list. I'm going to update the name and comment and remove the redundant `list()` calls on its return value.

rebecca-caktus commented on symposion/proposals/models.py:
```diff
@@ -109,19 +109,29 @@ class ProposalBase(models.Model):
+    def save(self, *args, **kwargs):
+        if self.pk:
+            # If already in database, update cached tags before saving
+            self.cached_tags = self.get_tags_display()
```
How does `cached_tags` get set if there is no `pk`?

dpoirier replied:
If there's no PK, there can't be any tags pointing at this record yet, so the default value `''` for cached_tags is correct.

rebecca-caktus commented on symposion/reviews/views.py:
```diff
@@ -263,6 +289,7 @@ def review_detail(request, pk):
 
                 tags = proposal_tags_form.cleaned_data['tags']
                 proposal.tags.set(*tags)
+                proposal.cache_tags()
```
I kind of wish that this were in a signal. Maybe the `save()` customization as well. What do you think?

dpoirier commented:
Updated to use signals; see if the code looks reasonable.

rebecca-caktus commented:
Looks great. I might suggest adding a few tests to assert the behavior of the `cache_tags` feature.

dpoirier commented:
Updated again.

rebecca-caktus commented:
Django docs suggest keeping signals in their own file & importing in the app's `ready()` function:

https://docs.djangoproject.com/en/1.8/topics/signals/ (see "Where should this code live?")

dpoirier commented:
On Thu, Aug 13, 2015 at 12:40 PM, Rebecca <notifications@github.com> wrote:

> Django docs suggest keeping signals in their own file & importing in the
> app's ready() function:
>
> https://docs.djangoproject.com/en/1.8/topics/signals/ (see "Where should
> this code live?")
>

​Thanks for keeping me up on the evolving Django best practices.  I've
moved the signals.

rebecca-caktus commented:
Django 1.8 strikes!

```
File "pycon/symposion/reviews/views.py" in proposals_list
  60.         for vote in LatestVote.objects.filter(user__pk=user_pk, proposal__in=queryset).select_related("vote")

Exception Type: FieldError at /2016/reviews/section/talks/
Exception Value: Non-relational field given in select_related: 'vote'. Choices are: proposal, user
```

rebecca-caktus commented:
After updating two `select_related` calls to `prefetch_related` locally, I'm still getting 1108 queries on `/2016/reviews/section/talks/`. Is this the best that can be done?

rebecca-caktus commented:
Ah, I'm seeing now that you got it to 21 before - so maybe something changed.

dpoirier commented:
I got the queries back down to 21 - see the commit message for an explanation.

rebecca-caktus commented:
Data in the "Category" column shows on staging, but not locally.

dpoirier commented:
Updated - category column should be fixed; optimized all the review views even more; added tests to make sure the category stays fixed and the views stay optimized.

rebecca-caktus commented:
I'm still seeing 561 queries here:

http://localhost:8000/2016/reviews/section/talks/

dpoirier commented:
There must be something different, obviously :-)  Maybe you have more
interesting local data.

I've been using the Django debug toolbar to see what queries occurred while
processing a request, specifically looking for queries where DDT notes that
it was repeated NN times for values of NN that look roughly like a multiple
of the number of talks.  We could get together and use that to see what's
going on when you load that page. Or you could give me a dump of your local
data and I could work with that.


*Dan Poirier*Developer

dpoirier@caktusgroup.com
www.caktusgroup.com

On Wed, Aug 26, 2015 at 4:52 PM, Rebecca <notifications@github.com> wrote:

> I'm still seeing 561 queries here:
>
> http://localhost:8000/2016/reviews/section/talks/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PyCon/pycon/pull/469#issuecomment-135167381>.
>

rebecca-caktus commented:
:+1:

hatlas commented:
Passed QA - https://us.pycon.org/2016/reviews/section/talks/ takes about 5 seconds to load, https://pycon-staging.global.ssl.fastly.net/2016/reviews/section/talks/ takes <1 second.
